pkg/controllermanager/controller/shootstate/add_test.go

```diff
@@ -0,0 +1,195 @@
+			Expect(requests).To(ConsistOf(reconciliationRequest))
+		})
+
+		It("should return nil if ShootState is not present", func() {
+			shoot := &gardencorev1beta1.Shoot{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      shootName,
+					Namespace: shootNamespace,
+				},
+			}
+			requests := reconciler.MapShootToShootState(log)(ctx, shoot)
+			Expect(requests).To(BeEmpty())
+		})
+
+		It("should return nil if argument is not a gardencorev1beta1.Shoot", func() {
+			deploymentName := "dep-1"
+			deploymentNamespace := "ns-1"
+			deployment := appsv1.Deployment{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      deploymentName,
+					Namespace: deploymentNamespace,
+				},
+			}
+
+			requests := reconciler.MapShootToShootState(log)(ctx, &deployment)
+			Expect(requests).To(BeNil())
+		})
+	})
+
+	Describe("ShootPredicate", func() {
+		var p predicate.Predicate
+
+		BeforeEach(func() {
+			p = reconciler.ShootPredicates()
+		})
+
+		Describe("#Update", func() {
+			var (
+				shoot    *gardencorev1beta1.Shoot
+				shootOld *gardencorev1beta1.Shoot
+
+				shootState *gardencorev1beta1.ShootState
+			)
+
+			BeforeEach(func() {
+				shoot = &gardencorev1beta1.Shoot{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      "shoot-",
+						Namespace: "ns-",
+					},
+					Status: gardencorev1beta1.ShootStatus{},
+				}
+				Expect(c.Create(ctx, shoot)).To(Succeed())
+
+				shootOld = shoot.DeepCopy()
+
+				shootState = &gardencorev1beta1.ShootState{
+					ObjectMeta: metav1.ObjectMeta{
+						Name:      shoot.Name,
+						Namespace: shoot.Namespace,
+					},
+				}
+				Expect(c.Create(ctx, shootState)).To(Succeed())
+			})
+
+			It("should return false because new object is not a shoot", func() {
+				Expect(p.Update(event.UpdateEvent{})).To(BeFalse())
+			})
+
+			It("should return false because old object is not a shoot", func() {
+				Expect(p.Update(event.UpdateEvent{ObjectNew: shoot})).To(BeFalse())
+			})
+
+			It("should return true if last operation transitions from Reconcile to Migrate", func() {
+				shootOld.Status.LastOperation = &gardencorev1beta1.LastOperation{
+					Type: gardencorev1beta1.LastOperationTypeReconcile,
+				}
+				shoot.Status.LastOperation = &gardencorev1beta1.LastOperation{
+					Type: gardencorev1beta1.LastOperationTypeMigrate,
+				}
+				Expect(p.Update(event.UpdateEvent{ObjectOld: shootOld, ObjectNew: shoot})).To(BeTrue())
+			})
+
+			It("should return true if last operation transitions from Migrate to Restore", func() {
+				shootOld.Status.LastOperation = &gardencorev1beta1.LastOperation{
+					Type: gardencorev1beta1.LastOperationTypeMigrate,
+				}
+				shoot.Status.LastOperation = &gardencorev1beta1.LastOperation{
+					Type: gardencorev1beta1.LastOperationTypeRestore,
+				}
+				Expect(p.Update(event.UpdateEvent{ObjectOld: shootOld, ObjectNew: shoot})).To(BeTrue())
+			})
+
+			It("should return true if Restore operation succeeds ", func() {
```

Comment:
Let's add one more case if the lastOperation is Restore/Succeeded and the new operation is also Restore/Succeeded.

Reply:
Thanks for the advice, included an additional test catching the case when the Restore operation has already succeeded.